[Community Pipeline] UnCLIP Text Interpolation Pipeline

Abhinay1997 opened the pull request.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/diffusers/pr_2257). All of your documentation changes will be reflected on that endpoint.

patrickvonplaten reviewed (approved):
This looks super cool, thanks for adding it @Abhinay1997 

Would be interesting to try it out on some of the examples of DALLE-2, e.g.

"a photo of an adult lion → a photo of lion cub"
"a photo of a landscape in winter → a photo of a landscape in fall"
"a photo of a victorian house → a photo of a modern house"

Also think this is a cool pipeline to build a spaces with :-) 

@williamberman can you also have a look here?

Abhinay1997 commented:
Thanks @patrickvonplaten. 😄

Need your input on the attention mask to be used for the interpolated text embeddings because the results are not great when the difference in prompt length is large.

patrickvonplaten commented:
> Thanks @patrickvonplaten. smile
> 
> Need your input on the attention mask to be used for the interpolated text embeddings because the results are not great when the difference in prompt length is large.

cc @williamberman maybe?

Abhinay1997 commented:
@patrickvonplaten , see our discussion here #1869 willberman suggested we use the larger of the two for now

williamberman commented on examples/community/unclip_text_interpolation.py:
```diff
@@ -0,0 +1,673 @@
+
+        # Interpolate from the start to end prompt using slerp and add the generated images to an image output pipeline
+        out = []
+        for interp_val in np.linspace(0, 1, steps):
+            # Use the start and end prompts for 0 and 1 values as slerp results are subjectively worse than slerp results for the same.
+            if interp_val == 0:
+                text_embeds = start_text_embeds
+                last_hidden_state = start_last_hidden_state
+            elif interp_val == 1:
+                text_embeds = end_text_embeds
+                last_hidden_state = end_last_hidden_state
+            else:
+                text_embeds = UnCLIPTextInterpolationPipeline.slerp(interp_val, start_text_embeds, end_text_embeds)
+                last_hidden_state = UnCLIPTextInterpolationPipeline.slerp(
+                    interp_val, start_last_hidden_state, end_last_hidden_state
+                )
+
+            text_model_output.text_embeds = text_embeds.unsqueeze(0).to(device)
+            text_model_output.last_hidden_state = last_hidden_state.unsqueeze(0).to(device)
+
+            res = self._generate(
+                text_model_output=text_model_output,
+                text_attention_mask=attention_mask,
+                generator=generator,
+                prior_num_inference_steps=prior_num_inference_steps,
+                decoder_num_inference_steps=decoder_num_inference_steps,
+                super_res_num_inference_steps=super_res_num_inference_steps,
+                prior_guidance_scale=prior_guidance_scale,
+                decoder_guidance_scale=decoder_guidance_scale,
+                output_type=output_type,
+                return_dict=return_dict,
+            )
+
```
I think ideally we should batch the embeddings instead of effectively running the pipeline in a loop

Abhinay1997 replied:
Sure. Will batch the pipeline run.

williamberman commented on examples/community/unclip_text_interpolation.py:
```diff
@@ -0,0 +1,673 @@
+                    interp_val, start_last_hidden_state, end_last_hidden_state
+                )
+
+            text_model_output.text_embeds = text_embeds.unsqueeze(0).to(device)
+            text_model_output.last_hidden_state = last_hidden_state.unsqueeze(0).to(device)
```
ideally we use the interpolated results directly instead of mutating `text_model_output`

Abhinay1997 replied:
Got it. Will make the change.

williamberman replied:
> @williamberman , made changes based on your feedback. Could you review them when you can ?
> 
> P.S Ran the code through black and isort multiple times but it's still failing the code quality test

We recently updated the versions of our linters etc.. could you try making sure they're up to date and running `make style` locally before pushing?

williamberman commented on examples/community/unclip_text_interpolation.py:
```diff
@@ -0,0 +1,673 @@
+        return ImagePipelineOutput(images=image)
+
+    @staticmethod
+    def slerp(val, low, high):
```
nice! slerp doesn't have to be a static or regular method on the class. Let's just move it to a regular function at the top of the file :)

Abhinay1997 replied:
Ohh. Yeah that makes sense.

williamberman commented on examples/community/unclip_text_interpolation.py:
```diff
@@ -0,0 +1,673 @@
+
+    @torch.no_grad()
+    # Copied from diffusers.pipelines.unclip.pipeline_unclip.UnCLIPPipeline.__call__
+    def _generate(
```
We try to keep the `__call__` function pretty self contained so lets move `_generate` back to inside `__call__`. This should work well with the other comment on batching the interpolated text embeddings :)

Abhinay1997 replied:
Do you mean like this ?

```python
def __call__(.....):
     def _generate(.....):
```

williamberman replied:
almost! could we just remove the `_generate` function and have all of the logic directly in the `__call__` method?

Abhinay1997 replied:
Sure !

williamberman commented:
Great start @Abhinay1997 !

Abhinay1997 commented:
@williamberman , made changes based on your feedback. Could you review them when you can ?

williamberman commented on examples/community/unclip_text_interpolation.py:
```diff
@@ -0,0 +1,666 @@
+        batch_text_embeds = []
+        batch_last_hidden_state = []
+
+        for interp_val in torch.linspace(0, 1, steps):
+            text_embeds = slerp(interp_val, text_model_output.text_embeds[0], text_model_output.text_embeds[1])
+            last_hidden_state = slerp(
+                interp_val, text_model_output.last_hidden_state[0], text_model_output.last_hidden_state[1]
+            )
+            batch_text_embeds.append(text_embeds.unsqueeze(0))
+            batch_last_hidden_state.append(last_hidden_state.unsqueeze(0))
+
+        batch_text_embeds = torch.cat(batch_text_embeds)
+        batch_last_hidden_state = torch.cat(batch_last_hidden_state)
```
nice!

williamberman commented:
Love the progress @Abhinay1997 ! Could you also add some example code for running the pipeline along with the outputs it gives :) ?

Abhinay1997 commented on examples/community/README.md:
````diff
@@ -951,3 +953,39 @@ print(pipeline.prior_scheduler)
 
 ![shiba-inu](https://user-images.githubusercontent.com/16448529/209185639-6e5ec794-ce9d-4883-aa29-bd6852a2abad.jpg)
 
+### UnCLIP Text Interpolation Pipeline
+
+This Diffusion Pipeline takes two prompts and interpolates between the two input prompts using spherical interpolation ( slerp ). The input prompts are converted to text embeddings by the pipeline's text_encoder and the interpolation is done on the resulting text_embeddings over the number of steps specified. Defaults to 5 steps. 
+
+```python
+import torch
+from diffusers import DiffusionPipeline
+
+device = torch.device("cpu" if not torch.cuda.is_available() else "cuda")
+
+pipe = DiffusionPipeline.from_pretrained(
+    "kakaobrain/karlo-v1-alpha",
+    torch_dtype=torch.float16,
+    custom_pipeline="unclip_text_interpolation"
+)
+pipe.to(device)
+
+start_prompt = "A photograph of an adult lion"
+end_prompt = "A photograph of a lion cub"
+#For best results keep the prompts close in length to each other. Of course, feel free to try out with differing lengths.
+generator = torch.Generator(device=device).manual_seed(42)
+
+output = pipe(start_prompt, end_prompt, steps = 6, generator = generator, enable_sequential_cpu_offload=False)
+
+for i,image in enumerate(output.images):
+    img.save('result%s.jpg' % i)
+```
+
+The resulting images in order:-
+
+![result_0](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_0.png)
+![result_1](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_1.png)
+![result_2](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_2.png)
+![result_3](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_3.png)
+![result_4](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_4.png)
+![result_5](https://huggingface.co/datasets/NagaSaiAbhinay/UnCLIPTextInterpolationSamples/resolve/main/lion_to_cub_5.png)
````
@williamberman Code example for the pipeline.

Abhinay1997 commented:
@williamberman 

1. Refactored the __call__ function as suggested :) 
2. I updated to the same linter versions as the github action and `make style` works on my local machine but fails the check for the PR.

williamberman commented on examples/community/unclip_text_interpolation.py:
```diff
@@ -0,0 +1,573 @@
+import inspect
+from typing import List, Optional, Tuple, Union
+
+import torch
+from torch.nn import functional as F
+
+from diffusers import (
+    DiffusionPipeline,
+    ImagePipelineOutput,
+    PriorTransformer,
+    UnCLIPScheduler,
+    UNet2DConditionModel,
+    UNet2DModel,
+)
+from diffusers.pipelines.unclip import UnCLIPTextProjModel
+from diffusers.utils import is_accelerate_available, logging, randn_tensor
+from transformers import CLIPTextModelWithProjection, CLIPTokenizer
+from transformers.models.clip.modeling_clip import CLIPTextModelOutput
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
```
```suggestion
import torch
from torch.nn import functional as F
from transformers import CLIPTextModelWithProjection, CLIPTokenizer
from transformers.models.clip.modeling_clip import CLIPTextModelOutput

from diffusers import (
    DiffusionPipeline,
    ImagePipelineOutput,
    PriorTransformer,
    UnCLIPScheduler,
    UNet2DConditionModel,
    UNet2DModel,
)
from diffusers.pipelines.unclip import UnCLIPTextProjModel
from diffusers.utils import is_accelerate_available, logging, randn_tensor


logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
```

williamberman replied:
Should help it pass CI :)

williamberman commented:
Awesome, looks basically good to do @Abhinay1997! I needed to merge in master to get the updated linter versions :)

Abhinay1997 commented:
Thanks for the help Will ! Hope we are good for the merge now.

williamberman commented:
Awesome, this is great @Abhinay1997! Would you be interested in making a spaces to showcase the pipeline? https://huggingface.co/spaces

Abhinay1997 commented:
Sure @williamberman ! I was thinking of doing it once the PR is merged :)
